### ISSUE TITLE
Services get stuck when node label is not defined

### ISSUE THREAD

**Issue by mattdavenport** [kontena/kontena#1275]
If a service is defined in `kontena.yml` with affinities such as `label==db` and there are no nodes with such label, the service gets stuck in a deployment state. It seems that this will randomly block further deployment steps, but sometimes deployment will continue leaving instances at 0 for the particular service. However there is no helpful message printed during deployment stating there is no node matching the affinity for the service.

**Comment by mattdavenport** [comment]
Furthermore, app logs for the service provide no output, and logs from the `kontena_master` container only show constant GET requests to the `/containers` endpoint for the service.

**Comment by colinmollenhour** [comment]
It seems "deploy" is expecting some condition which doesn't need to happen. Using a config like:

```yaml
  scaling: 1
  strategy: daemon
  deploy_opts:
    min_health: 0.0
  affinity: 
    - label==db-node
```

with nodes:

```
$ k node ls
Name                                                                   Status     Initial    Labels                                  
test2.local.lan                                                        online     yes        region=packet-ewr             
test3.local.lan                                                        online     no         region=packet-ewr,db-node
test4.local.lan                                                        online     no         region=packet-ewr,db-node
test5.local.lan                                                        online     no         region=packet-ewr,db-node
```

I run the deploy to add a node (two already were running, added db-node label to third, ran deploy). It spins for a long time and finally times out:

```
$ k app deploy galera-node
 [done] Updating galera-node      
 [fail] Deploying galera-node      
 [error] deploy timed out
```

However, while it was spinning the master was getting hit with requests almost every second:

```
"GET /v1/services/mytest/magento-galera-node/containers HTTP/1.1" 200 6539 0.1671
"GET /v1/services/mytest/magento-galera-node/containers HTTP/1.1" 200 6539 0.1382
"GET /v1/services/mytest/magento-galera-node/containers HTTP/1.1" 200 6539 0.1372
"GET /v1/services/mytest/magento-galera-node/containers HTTP/1.1" 200 6539 0.1378
```

Checking the containers via CLI *while the deploy was still spinning* I got:

```
$ k service containers magento-galera-node
mytest/magento-galera-node/magento-galera-node-1:
  rev: 2016-11-04 17:23:22 UTC
  node: test4.local.lan
  dns: magento-galera-node-1.mytest.kontena.local
  ip: 10.81.1.165
  public ip: 147.75.***.***
  status: running

mytest/magento-galera-node/magento-galera-node-2:
  rev: 2016-11-04 17:23:22 UTC
  node: test5.local.lan
  dns: magento-galera-node-2.mytest.kontena.local
  ip: 10.81.1.166
  public ip: 147.75.***.***
  status: running

mytest/magento-galera-node/magento-galera-node-3:
  rev: 2016-11-04 17:23:22 UTC
  node: test3.local.lan
  dns: magento-galera-node-3.mytest.kontena.local
  ip: 10.81.1.167
  public ip: 147.75.***.***
  status: running
```

So the containers were all running on the nodes as expected but the deploy command seemed to think the deploy wasn't complete..

**Comment by colinmollenhour** [comment]
Also it seems that the tools should show "3/3" instead of "3/1" in cases where the "daemon" strategy is used. In the example above it seems this should be easy to figure out:

expected = (nodes matching affinities) * (scaling)

```
$ k app ls
NAME                           IMAGE                                              INSTANCES       STATEFUL   STATE           
galera-seed                    registry.kontena.local/mariadb-galera:latest       0 / 1           yes        stopped
galera-node                    registry.kontena.local/mariadb-galera:latest       3 / 1           yes        running
galera-nowsrep                 registry.kontena.local/mariadb-galera:latest       0 / 1           yes        stopped
```

You can see there are also some 0/1. These are services which no longer match any nodes because I removed the labels that are required in the affinities (seed was used to bootstrap cluster, nowsrep was used to run "mysql_upgrade" and reset root password if needed).

**Comment by jnummelin** [comment]
The CLI is basically just blindly waiting for specific amount of instances (containers) to be running and is happy only if it sees those. :) The referenced PR fixes this by having better status info of the actual deployment and also tracks the reason why deployment failed if it did. It still requires some fine tuning, but should get shipped with 0.17 release.

**Comment by jnummelin** [comment]
closed via #1348
